src/OmniSharp.Roslyn.CSharp/Services/Structure/BlockStructureService.cs

```diff
@@ -50,7 +53,15 @@ public BlockStructureService(IAssemblyLoader loader, OmniSharpWorkspace workspac
 
         public async Task<BlockStructureResponse> Handle(BlockStructureRequest request)
         {
+            // Waiting until the document is fully formed in memory (for project systems that have this ability) 
+            // helps to reduce chances of returning invalid document block structure while compilation is still in progress.
+            await _projectSystems.WaitForAllProjectsToLoadForFileAsync(request.FileName);
```

Comment:
This doesn't seem like the right abstraction. 
If I write something that uses OmniSharp workspace, but I don't import all the project systems to call this API, what happens?

If a requirement of GetDocument is that we wait for the project systems, we should move this into GetDocument (or some abstraction over the combination thereof)

Reply:
Case in point: BlockStructure is a feature that only require the syntax tree for a document. It doesn't do any cross/document cross/project analysis...

---
In reply to: [232089084](https://github.com/OmniSharp/omnisharp-roslyn/pull/1322#discussion_r232089084) [](ancestors = 232089084)

Reply:
The Visual Studio components of Roslyn have a term for this: you can request a document with "full" or "partial" semantics. Requesting partial semantics will result in less work being done to produce the document + project graph.

---
In reply to: [232089407](https://github.com/OmniSharp/omnisharp-roslyn/pull/1322#discussion_r232089407) [](ancestors = 232089407,232089084)

Reply:
@rchande : Very good points! Indeed introducing the concept of a partial vs fully loaded semantic model would be a better abstraction. At the moment the behavior of existing OmniSharpWorkspace.GetDocumentXXXX methods IMO matches more to retrieving data from a partial semantic model. The alternative design that I would like to propose (to the one that is in this PR) is to add methods like the following. The methods will be waiting until all projects are loaded before returning the result:

```
        public DocumentId GetDocumentIdFromFullSemanticModel(string filePath)
        public IEnumerable<Document> GetDocumentsFromFullSemanticModel(string filePath)
        public Document GetDocumentFromFullSemanticModel(string filePath)
```

Regarding BlockStructure requests processing – it turned out that they do need to use the full semantic model because otherwise blocks like doc-comments or usings that might not be marked as collapsible in the editor.

Reply:
@dmgonch I agree with the concept but think we can improve on the terminology at little. Instead of `SemanticModel`, maybe we should be explicit about how this waits for all the projects to finish? `GetDocument*FromFullProjectModel`?

Reply:
@rchande : `GetDocument*FromFullProjectModel` sounds better - I will go with it. Thanks for the feedback!

Reply:
@rchande Reworked and retested changes as discussed above. The new methods are Async because they need to perform wait while projects are being loaded and request handlers calling the methods are already async as well.